flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/restore/ChangelogBackendRestoreOperation.java

```diff
@@ -80,14 +89,19 @@
             ChangelogKeyedStateBackend<?> backend,
             ChangelogStateBackendHandle backendHandle,
             StateChangelogHandleReader<T> changelogHandleReader,
-            ClassLoader classLoader)
+            ClassLoader classLoader,
+            Map<UUID, Map<Short, StateID>> metadataByBackend)
             throws Exception {
         for (ChangelogStateHandle changelogHandle :
                 backendHandle.getNonMaterializedStateHandles()) {
+            Map<Short, StateID> stateIds =
+                    metadataByBackend.computeIfAbsent(
+                            backendHandle.getBackendID(), ign -> new HashMap<>());
```

Comment:
To answer the question that whether to add a unit test, I checked all tests under `flink-statatebackend-changelog` module, and found that the `stateIds` here is always empty. This could be explained that we would not restore two state handles from previous same statebackend, and the statd-ids are updated during `applyOperation`.
If so, do we still need the API of `getBackendUID`?

Reply:
You are right, I've removed backend ID (both from handle and from backend itself).

I started to [refactor](https://github.com/rkhachatryan/flink/tree/refactor-backend-test) rescaling tests to add more snapshots for the same backend and then shuffle them (so that metadata from one backend could overwrite metadata of another).

But I realized that it won't break recovery, because each snapshot (i.e. `ChangelogStateBackendHandle`) contains metadata in the beginning; and reading of each such snapshot is "atomic" (i.e. doesn't interlieave with reading other snapshots).